### PR TITLE
feat(robot-server): add maintenance run resource and routes

### PR DESCRIPTION
**Pull request by sanni-t**
Closes RSS-199

# Overview

Adds a `/maintenance_runs` resource to run maintenance procedures like LPC and instrument attach/detach/calibration.

[This PR is taking over from previous #12462 to resolve conflicts from multi-branching. Review comments from that PR will be addressed here]

# Test Plan

- [ ] Use the `/maintenance_runs` endpoints to create a maintenance run, issue commands to the run, delete the run. 
- [ ]  Create protocol run while a maintenance run is active, it should delete the maintenance run and create the protocol run successfully.
- [ ] Send a new maintenance run creation request while a maintenance run is already active; the old one should get replaced by the new one.
- [ ] Start running a protocol run and issue a maintenance run creation request; it should throw an error and not allow creating the maintenance run.
- [ ] While a maintenance run is active when a protocol run is idle, issue a PLAY action on the protocol run; the maintenance run should get removed and the protocol run should start playing.
- [ ] Test uploading a protocol, running LPC and running the protocol via the app. The changes in this PR should not affect any existing behavior of the `/runs` resource.

# Changelog

- Adds a new maintenance runs resource that implements a base router, a commands router and a labware router over `/maintenance_runs` route
- Adds supporting classes of `MaintenanceEnginestore` and `MaintenanceRunDataManager` that are very similar to the ones used by the `/runs` resource but without the run store and protocol management.
- Adds maintenance run deletion when `PLAY` action is issued over `runs` endpoint.

# Review requests

- Make sure code looks good and has good unit test coverage
- Make sure the above tests in the plan succeed as expected
- The intention of approach we took in this PR - duplication of a lot of code between the runs & maintenance runs resource- was done on purpose to do a first pass of run split without affecting any code related to protocol runs. We expected that we would eventually be able to compress some of the code into a shared resource and I think this PR proves that we can, and we should, but I propose we work on that refactor in a follow-up PR. This PR establishes a base for splitting the run and builds client-server expectations on managing two separate runs. The follow-up will work on code-management and finer details of the client-server interactions while using such separate runs.

Follow-up tickets:

- RSS-223
- RSS-224
- RSS-225

# Risk assessment

Low-None for existing infrastructure since this adds a new resource that isn't used by anything yet.

### CODE REVIEW

**Review by SyntaxColoring — APPROVED** [review]
Approved previously by others on #12462.

Discussed with @b-cooper, @jbleon95, @smb2268 and @TamarZanzouri in-person. We should merge this now if all comments so far have been addressed to @sanni-t's satisfaction, because we want to unblock app development and optimistically make the internal testing release cutoff.

I still have a pending review, which I'll post here. It shouldn't block merging this now.

**Review by b-cooper — APPROVED** [review]
🦾

File: robot-server/tests/maintenance_runs/router/test_base_router.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
+"""Tests for base /runs routes."""
```
Comment (SyntaxColoring) [inline robot-server/tests/maintenance_runs/router/test_base_router.py:1]: I think we also need integration tests for the maintenance runs API. Specifically, we need integration tests to cover the interplay between `/runs` and `/maintenance_runs` and how they sometimes conflict, sometimes overlap, and sometimes automatically clobber each other.

The unit tests don't seem sufficient to me because multiple layers contribute to the API's behavior.

I think this is easiest if we do it soon, because the intended behavior is kind of complicated, so it's easier if we do it while it's fresh in our minds.

File: robot-server/robot_server/maintenance_runs/router/base_router.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    path="/maintenance_runs",
+    summary="Create a maintenance run",
+    description=dedent(
+        """
+        Create a new maintenance run to track robot interaction.
+
+        If a maintenance run already exists, it will be cleared
+        and a new one will be created.
+
+        Will raise an error if a *protocol* run exists and is not idle.
```
Comment (SyntaxColoring) [inline robot-server/robot_server/maintenance_runs/router/base_router.py:129]: Continuing the terminology question that you brought up about what to call the "other kind" of run, I guess we have two options. Do you have a preference?

**Option 1:**

Avoid saying "protocol run," because it can unintentionally sound like it doesn't apply to resources under `/runs` where `protocolId` is null. So in this case, we would have to say something like:

> Will raise an error if a run from `/runs` exists and is not idle.

**Option 2:**

Lean into calling `/runs` "protocol runs," even though "protocol" isn't mentioned in their route URLs, and they sometimes have `"protocolId": null`. Update the documentation and our internal class names to say "protocol runs" where we mean `/runs`, "maintenance runs" where we mean `/maintenance_runs`, and "runs" where we mean either.

File: robot-server/robot_server/router.py [hunk 3]
```diff
@@ -48,6 +49,12 @@
     dependencies=[Depends(check_version_header)],
 )
 
+router.include_router(
+    router=maintenance_runs_router,
+    tags=["Maintenance Run Management"],
+    dependencies=[Depends(check_version_header)],
```
Comment (SyntaxColoring) [inline robot-server/robot_server/router.py:55]: Do you have an opinion for whether these new endpoints to require a newer `Opentrons-Version` header value? I honestly don't know.

File: robot-server/robot_server/runs/dependencies.py [hunk 4]
```diff
@@ -55,6 +55,17 @@ async def get_engine_store(
...
 
 
+async def get_protocol_run_has_been_played(
+    engine_store: EngineStore = Depends(get_engine_store),
+) -> bool:
+    """Whether the current protocol run, if any, has been played."""
+    try:
+        protocol_run_state = engine_store.engine.state_view
+    except AssertionError:
+        return False
```
Comment (SyntaxColoring) [inline robot-server/robot_server/runs/dependencies.py:65]: Can we change this to catch a specific exception that's part of `EngineStore`'s public interface?

File: robot-server/robot_server/maintenance_runs/maintenance_action_models.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+"""Request and response models for controlling maintenance runs with actions."""
+from datetime import datetime
+from enum import Enum
+from pydantic import BaseModel, Field
+
+from robot_server.service.json_api import ResourceModel
+
+
+class MaintenanceRunActionType(str, Enum):
```
Comment (SyntaxColoring) [inline robot-server/robot_server/maintenance_runs/maintenance_action_models.py:9]: I've lost track of our short- and long-term plans for actions in `/maintenance_runs`, but if they're going to exist like this, it seems like they can be deduplicated with the models for `/runs`?

File: robot-server/robot_server/maintenance_runs/__init__.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+"""Maintenance Run creation and management.
+
+A "run" is a logical container for a user's interaction with a robot,
+usually (but not always) with a well-defined start point and end point.
+
+A maintenance run is a special type of run that is used for running
+maintenance procedures like instrument attach/detach/calibration and LPC.
+A maintenance run doesn't have a protocol associated with it, but is issued individual
+commands and actions over HTTP.
+"""
```
Comment (SyntaxColoring) [inline robot-server/robot_server/maintenance_runs/__init__.py:10]: I remain very concerned about this HTTP API approach. This ship has long since sailed, and my concerns extend beyond the scope of this PR, but I want to note them here for the record. I've previously brought these up individually, scattered across RSS↔RAUT syncs and in RSS calls.

* A route-level split between `/runs` and `/maintenance_runs` doubles the surface area of our run API, making it more difficult to document and explain. Especially because these two resources sometimes behave the same and sometimes behave differently.
* Having a `/maintenance_runs` resource that's parallel to `/runs` creates conceptual and terminological confusion everywhere about what we mean by "a run."
* The interplays between `/runs`↔`/runs`, `/runs`↔`/maintenance_runs`, and `/maintenance_runs`↔`/maintenance_runs` are nuanced and confusing. Sometimes these things raise conflict errors, sometimes they coexist concurrently, and sometimes they delete or un-current each other automatically. The fact that a `/run` can coexist at the same time as a `/maintenance_run` at all is confusing to me, from a perspective of only one thing can be controlling the robot at once. Only one should be "active," but it's difficult to correctly define "active."
* Aside from multiple runs coexisting, the ways that `/maintenance_runs` differs from `/runs`—auto-play, whether `pause` is allowed, and whether they persist across reboots—are all orthogonal to each other, and defining a whole new resource type for them seems unnecessary when we could have added more parametrization to `/runs` like we already have with `protocolId`. This specific combination of behaviors is unnecessarily coupled to the way that the Opentrons App does things today.

The problem that we're solving with this PR is that Labware Position Check shares the same `ProtocolEngine` as the protocol itself, which causes all kinds of state bleed problems today. And we couldn't let that get worse when the Flex uses `ProtocolEngine` to do things like pipette attachment. I think we had ways to fix that problem that would have mitigated the concerns above, and wouldn't have been significantly more work for us today.

-----

Edit to clarify:

None of this is your fault, or new to this PR. This PR is the faithful implementation of much-discussed architectural decisions that were made with all of our participation.

File: robot-server/robot_server/maintenance_runs/maintenance_engine_store.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+class RunnerEnginePair(NamedTuple):
+    """A stored ProtocolRunner/ProtocolEngine pair."""
+
+    run_id: str
+    created_at: datetime
+    runner: LiveRunner
+    engine: ProtocolEngine
+
+
+class MaintenanceEngineStore:
```
Comment (SyntaxColoring) [inline robot-server/robot_server/maintenance_runs/maintenance_engine_store.py:36]: Given the interplay between `/runs` and `/maintenance_runs`, and how they sometimes overlap, sometimes conflict-error, and sometimes automatically clobber each other, would it be easier if they shared a single store instead of having a separate `MaintenanceEngineStore` and `EngineStore`?

The robot server is responsible for enforcing a very important mutual exclusion property: only one run or maintenance run can interact with the hardware controller at a time. This seems infeasible to maintain when our enforcement has to be spread across several router functions: `POST /runs`, `POST /maintenance_runs`, `POST /runs/actions`, and `POST /maintenance_runs/actions`, at least.

A single store would centralize the logic of when you're allowed to `POST` a new run or maintenance run. It wouldn't necessarily help with when you're allowed to `POST` a new command or play it, though. I'm not sure what to do about that. Maybe we should do the same thing with `RunController`.

File: robot-server/robot_server/maintenance_runs/router/base_router.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+@base_router.post(
+    path="/maintenance_runs",
+    summary="Create a maintenance run",
+    description=dedent(
+        """
+        Create a new maintenance run to track robot interaction.
+
+        If a maintenance run already exists, it will be cleared
+        and a new one will be created.
```
Comment (SyntaxColoring) [inline robot-server/robot_server/maintenance_runs/router/base_router.py:127]: What happens (or should happen) if a maintenance run exists and is actively running a command? Does this `POST` wait for the current command to complete, or does it wait for all commands to complete, or does it interrupt the command as if we issued a stop action, or does it raise a conflict error?

File: robot-server/robot_server/runs/router/actions_router.py [hunk 9]
```diff
@@ -78,20 +84,34 @@ async def create_run_action(
...
     created_at: datetime = Depends(get_current_time),
+    maintenance_engine_store: MaintenanceEngineStore = Depends(
+        get_maintenance_engine_store
+    ),
 ) -> PydanticResponse[SimpleBody[RunAction]]:
     """Create a run control action.
 
+    When a play action is issued to a protocol run while a maintenance run is active,
+    the protocol run is given priority and the maintenance run is deleted before
+    executing the protocol run play action..
```
Comment (SyntaxColoring) [inline robot-server/robot_server/runs/router/actions_router.py:95]: Same question as my other note. What happens (or should happen) if a maintenance run exists and is actively running a command?

File: robot-server/robot_server/runs/router/commands_router.py
Comment (SyntaxColoring) [inline robot-server/robot_server/runs/router/commands_router.py:1]: Something should probably prevent you from using `"intent": "setup"` commands with `/runs` if you're also using `/maintenance_runs`, and vice-versa, right? `"intent": "setup"` commands interacting with maintenance runs seems infeasible for anyone to reason about, and it should never be necessary. (I desperately hope, anyway?)

File: robot-server/tests/maintenance_runs/router/test_base_router.py (hunk 1 above)
Comment (TamarZanzouri) [reply robot-server/tests/maintenance_runs/router/test_base_router.py:1]: I agree. We have integration tests but they are very straightforward. We can test all these edge cases in the integration tests.

Comment (sanni-t) [reply robot-server/tests/maintenance_runs/router/test_base_router.py:1]: Yep, it's one of the follow-up tickets linked in the description above.

File: robot-server/robot_server/router.py (hunk 3 above)
Comment (sanni-t) [reply robot-server/robot_server/router.py:55]: Ya, I am not sure either. We should check the pros and cons doing that. The only one I can think of is being able to handle client-server version mismatch more greacefully. But we are anyway requiring the app to check the server version before starting a connection so there doesn't seem to be much of a gain (except for maybe third-party clients?)

File: robot-server/robot_server/maintenance_runs/maintenance_action_models.py (hunk 5 above)
Comment (sanni-t) [reply robot-server/robot_server/maintenance_runs/maintenance_action_models.py:9]: Yep, I plan on re-evaluating whether these models can also be deduplicated with the ones in `/runs`, just like error models

File: robot-server/robot_server/runs/router/commands_router.py
Comment (sanni-t) [reply robot-server/robot_server/runs/router/commands_router.py:1]: That's a good point. Originally the plan was to remove setup intent altogether so this situation would never arise, but now that we're going to deprecate it instead, we do need to have something that prevents the use of setup intent in /runs with a maintenance run active. 
Added RSS-230 to handle this.
